lib/internal/errors.js

```diff
@@ -0,0 +1,258 @@
+'use strict';
+
+// The whole point behind this internal module is to allow Node.js to no
+// longer be forced to treat every error message change as a semver-major
+// change. The NodeError classes here all expose a `code` property whose
+// value identifies the error.
+
+const kCode = Symbol('code');
+
+class NodeError extends Error {
+  constructor(key, ...args) {
+    super(message(key, args));
+    this[kCode] = key;
+    Error.captureStackTrace(this, NodeError);
+  }
+
+  get name() {
+    return `Error[${this[kCode]}]`;
+  }
+
+  get code() {
+    return this[kCode];
+  }
+}
+
+class NodeTypeError extends TypeError {
+  constructor(key, ...args) {
+    super(message(key, args));
+    this[kCode] = key;
+    Error.captureStackTrace(this, NodeTypeError);
+  }
+
+  get name() {
+    return `TypeError[${this[kCode]}]`;
+  }
+
+  get code() {
+    return this[kCode];
+  }
+}
+
+class NodeRangeError extends RangeError {
+  constructor(key, ... args) {
+    super(message(key, args));
+    this[kCode] = key;
+    Error.captureStackTrace(this, NodeRangeError);
+  }
+
+  get name() {
+    return `RangeError[${this[kCode]}]`;
+  }
+
+  get code() {
+    return this[kCode];
+  }
+}
+
+module.exports.Error = NodeError;
+module.exports.TypeError = NodeTypeError;
+module.exports.RangeError = NodeRangeError;
+
+var assert, util;
+function lazyAssert() {
+  if (!assert)
+    assert = require('assert');
+  return assert;
+}
+
+function lazyUtil() {
+  if (!util)
+    util = require('util');
+  return util;
+}
+
+function message(key, args) {
+  const assert = lazyAssert();
+  const util = lazyUtil();
+  const msg = Error[Symbol.for(key)];
+  // We want to assert here because if this happens,
+  // Node.js is doing something incorrect.
+  assert(msg, `An invalid error message key was used: ${key}.`);
+  let fmt = util.format;
+  if (typeof msg === 'function') {
+    fmt = msg;
+  } else {
+    args.unshift(msg);
+  }
+  const res = fmt.apply(null, args);
+  // We want to assert here because if this happens,
+  // Node.js is doing something incorrect.
+  assert(res, 'An invalid formatted error message was returned.');
+  return res;
+}
+
+// Below is the catalog of messages
+// Error keys should be all caps, value can
+// be a string formatted for use with
+// util.format, or a function.
+
+// Utility function for registering the error codes.
+function E(sym, val) {
```

Comment:
It could but doubt there'd be significant value in doing so. This is only used within this one module where we can easily ensure that the symbol values are always caps.